### PR TITLE
Fixing example compilation on esp-idf v5

### PR DESCRIPTION
**Pull request by davidswarbrick**
Currently a draft PR, addressing #4 (example failing to build). 
This removes legacy esp-idf example code and updates to compile against the latest stable esp-idf (v5).

### CODE REVIEW

**Comment by davidswarbrick** [comment]
After this commit, the example compiles, with warnings regarding the deprecated `driver/i2s.h`. The functionality is currently untested, once I've tested it on my Puca DSP and confirmed working it'll be ready for review/merge. The large line diff is due to sdkconfig changes which occur automatically when building against v5 of the esp-idf.

**Comment by davidswarbrick** [comment]
The example now compiles and runs using the v5 esp-idf tools:
- Compiles with `idf.py build`
- Flashes with `idf.py flash`
- Output of `idf.py monitor` shown below.

An audible sinewave was heard on the audio output as expected.

```
--- idf_monitor on /dev/ttyUSB0 115200 ---
--- Quit: Ctrl+] | Menu: Ctrl+T | Help: Ctrl+T followed by Ctrl+H ---
ets Jun  8 2016 00:22:57

rst:0x1 (POWERON_RESET),boot:0x13 (SPI_FAST_FLASH_BOOT)
configsip: 188777542, SPIWP:0xee
clk_drv:0x00,q_drv:0x00,d_drv:0x00,cs0_drv:0x00,hd_drv:0x00,wp_drv:0x00
mode:DIO, clock div:2
load:0x3fff0030,len:6972
ho 0 tail 12 room 4
load:0x40078000,len:15492
load:0x40080400,len:3844
0x40080400: _init at ??:?

entry 0x4008064c
I (30) boot: ESP-IDF v5.1-dev-2509-gcfef24863f 2nd stage bootloader
I (30) boot: compile time Dec 22 2022 23:26:43
I (31) boot: chip revision: v1.0
I (35) boot.esp32: SPI Speed      : 40MHz
I (40) boot.esp32: SPI Mode       : DIO
I (45) boot.esp32: SPI Flash Size : 2MB
I (49) boot: Enabling RNG early entropy source...
I (54) boot: Partition Table:
I (58) boot: ## Label            Usage          Type ST Offset   Length
I (65) boot:  0 nvs              WiFi data        01 02 00009000 00006000
I (73) boot:  1 phy_init         RF data          01 01 0000f000 00001000
I (80) boot:  2 factory          factory app      00 00 00010000 00100000
I (88) boot: End of partition table
I (92) esp_image: segment 0: paddr=00010020 vaddr=3f400020 size=0bbech ( 48108) map
I (118) esp_image: segment 1: paddr=0001bc14 vaddr=3ffb0000 size=01f9ch (  8092) load
I (121) esp_image: segment 2: paddr=0001dbb8 vaddr=40080000 size=02460h (  9312) load
I (128) esp_image: segment 3: paddr=00020020 vaddr=400d0020 size=1d8a4h (120996) map
I (176) esp_image: segment 4: paddr=0003d8cc vaddr=40082460 size=0af2ch ( 44844) load
I (201) boot: Loaded app from partition at offset 0x10000
I (201) boot: Disabling RNG early entropy source...
I (213) cpu_start: Pro cpu up.
I (213) cpu_start: Starting app cpu, entry point is 0x400812a8
0x400812a8: call_start_cpu1 at /home/david/esp/esp-idf/components/esp_system/port/cpu_start.c:149

I (0) cpu_start: App cpu up.
I (227) cpu_start: Pro cpu start user code
I (227) cpu_start: cpu freq: 160000000 Hz
I (227) cpu_start: Application information:
I (232) cpu_start: Project name:     esp32_picodsp_setup
I (238) cpu_start: App version:      3b1593a
I (243) cpu_start: Compile time:     Dec 30 2022 16:23:50
I (249) cpu_start: ELF file SHA256:  1fe9e208aa826810...
I (255) cpu_start: ESP-IDF:          v5.1-dev-2509-gcfef24863f
I (261) cpu_start: Min chip rev:     v0.0
I (266) cpu_start: Max chip rev:     v3.99
I (271) cpu_start: Chip rev:         v1.0
I (276) heap_init: Initializing. RAM available for dynamic allocation:
I (283) heap_init: At 3FFAE6E0 len 00001920 (6 KiB): DRAM
I (289) heap_init: At 3FFB28E0 len 0002D720 (181 KiB): DRAM
I (295) heap_init: At 3FFE0440 len 00003AE0 (14 KiB): D/IRAM
I (301) heap_init: At 3FFE4350 len 0001BCB0 (111 KiB): D/IRAM
I (308) heap_init: At 4008D38C len 00012C74 (75 KiB): IRAM
I (316) spi_flash: detected chip: gd
I (318) spi_flash: flash io: dio
W (322) spi_flash: Detected size(4096k) larger than the size in the binary image header(2048k). Using the size in the binary image header.
W (336) ADC: legacy driver is deprecated, please migrate to `esp_adc/adc_oneshot.h`
W (344) i2s(legacy): legacy i2s driver is deprecated, please migrate to use driver/i2s_std.h, driver/i2s_pdm.h or driver/i2s_tdm.h
I (357) app_start: Starting scheduler on CPU0
I (361) app_start: Starting scheduler on CPU1
I (361) main_task: Started on CPU0
I (371) main_task: Calling app_main()

 free mem=298768, written data=1280
I (911) main_task: Returned from app_main()
```

File: puca_dsp-esp-idf/esp-idf_puca_dsp_setup/main/Audio.cpp [hunk 1]
```diff
@@ -25,8 +25,11 @@ void Audio::I2S_init(void)
 
     i2s_driver_install(I2S_NUM_0, &i2s_config, 0, NULL);                            // install the driver on Core 0
     i2s_set_pin(I2S_NUM_0, &pin_config);                                            // port & pin config 
-    PIN_FUNC_SELECT(PERIPHS_IO_MUX_GPIO0_U, FUNC_GPIO0_CLK_OUT1);                   // MCLK on GPIO0
-    REG_WRITE(PIN_CTRL, 0xFFFFFFF0);                                                
+    // Deprecated pin setting for clk setup:
+    // PIN_FUNC_SELECT(PERIPHS_IO_MUX_GPIO0_U, FUNC_GPIO0_CLK_OUT1);                   // MCLK on GPIO0
+    // REG_WRITE(PIN_CTRL, 0xFFFFFFF0);                                                
+    uint32_t bits_cfg = (I2S_BITS_PER_CHAN_32BIT << 16) | I2S_BITS_PER_SAMPLE_16BIT;
+    i2s_set_clk(I2S_NUM_0, 22050, bits_cfg, I2S_CHANNEL_STEREO);
```
Comment (andy-wi) [inline puca_dsp-esp-idf/esp-idf_puca_dsp_setup/main/Audio.cpp]: In `i2s_set_clk`, `22050` should be `32000` (the sample rate in this example), or maybe better `I2S_SAMPLE_RATE` ?

Comment (davidswarbrick) [reply puca_dsp-esp-idf/esp-idf_puca_dsp_setup/main/Audio.cpp]: Yes good point! Have implemented this change.

**Comment by andy-wi** [comment]
Ah okay @davidswarbrick so the project files in this directory are PlatformIO projects, using the Espressif Platform and esp-idf framework inside PlatformIO, not direct ESP-IDF projects. This is maybe why you had some issue with the documentation in the earlier issue that you opened and then closed, sorry I only saw the thread a couple of hours ago. 

The latest ESP-IDF version supported in PlatformIO is currently v4, so, I've renamed the directories, added a readme for each directory, and also added a new directory for ESP-IDF projects [here](https://github.com/ohmic-net/puca_dsp/tree/main/puca_dsp-esp-idf). Could you add your example project for ESP-IDF v5 in this new directory?

**Comment by davidswarbrick** [comment]
Thanks for this @andy-wi  that makes a lot more sense! Yes, I'd be happy to add an example project. 
I could either add:
1. The code from this PR, which is really v4 code that now compiles against v5 (with lots of "this driver is deprecated" warnings). This would mean the example would be mostly similar to the v4 PlatformIO code, but will need to be rewritten when driver support is removed.
2. Or, I could add some code from my own [PÚCA DSP project](https://github.com/davidswarbrick/swarbs_dsp), which is currently a stripped down and updated version to make a minimal v5 implementation (i.e. with the new driver headers) with the test-tone re-implemented and audio passthrough from input -> output. I would add a line in the README to direct esp-idf v4 users to the PlatformIO example.

I expect option 2 may be more useful, but happy to do 1 if you feel that would make the v4 -> v5 transition any easier.

**Comment by andy-wi** [comment]
@davidswarbrick Great, thanks! 

Yes let's update this PR to leave the files in the [platformio-espidf directory](https://github.com/ohmic-net/puca_dsp/tree/main/puca_dsp-platformio-esp-idf) as is for now and they can be updated when PIO has v5 support and go with your option 2 above of adding a new example with a minimal v5 implementation in the [espidf directory](https://github.com/ohmic-net/puca_dsp/tree/main/puca_dsp-esp-idf).

**Comment by davidswarbrick** [comment]
Hi @andy-wi, I've added my minimal v5 implementation ready for a review, cheers :smile:

**Comment by andy-wi** [comment]
Thanks @davidswarbrick looks ok except for the sdkconfig file [here](puca_dsp-platformio-esp-idf/esp-idf_puca_dsp_mics_omnidirectional/sdkconfig) is edited also with the v5 sdkconfig so if you could push another commit to undo these changes and then I'll merge it?

**Comment by davidswarbrick** [comment]
@andy-wi your link seems to be broken, but looking at the URL you've tried to link I remember my old fork did have changes to the `mics_omnidirectional` example, but I've since removed those and the new content of this PR is only to do with files in `puca_dsp/puca_dsp-esp-idf`, as shown by the diff that Github provides. Perhaps your local copy of my fork isn't inline with Github? Do let me know if I'm mistaken and you'd still like a change.

**Comment by andy-wi** [comment]
@davidswarbrick thanks it wasn't, all good 🙂

**Pull request merged**